core/trino-main/src/main/java/io/trino/util/DateTimeUtils.java

```diff
@@ -67,9 +68,70 @@ public static int parseDate(String value)
         // - the required format is 'YYYY-MM-DD'
         // - all components should be unsigned numbers
         // https://github.com/trinodb/trino/issues/10677
+        Optional<Integer> days = tryParseDateYyyyMmDd(value);
+        if (days.isPresent()) {
+            return days.get();
+        }
         return toIntExact(TimeUnit.MILLISECONDS.toDays(DATE_FORMATTER.parseMillis(value)));
     }
 
+    /**
+     * Check if the date is in format yyyy-MM-dd, if yes, return the epoch days; otherwise
+     * return Optional.empty(), indicating the parse failed, and should fall back to generic
+     * parsing. Mixing the format checking and parsing in one pass for better performance.
+     * @param value date string to be parsed
+     * @return the number of days since 1970-01-01
+     */
+    public static Optional<Integer> tryParseDateYyyyMmDd(String value)
```

Comment:
How about using `java.util.OptionalInt` as a return type? it would make it easier for jit to optimize code (avoid boxed type allocation etc.).

Reply:
thanks for the suggestion. Updated to OptionalInt in 921b9c23e9894fdc57a581a29cb5d6aee89a99f9